### PR TITLE
Type hints for combine functions

### PR DESCRIPTION
**Pull request by TomNicholas**
Added type hints to `combine_nested` and `combine_by_coords`. 

Builds on #4696 because that PR generalised the argument types to include DataArrays, but I couldn't see that branch in the list to base this PR off of.

The "nested list-of-lists" argument to `combine_nested` opens up a can of worms: the only way I can see to specify the type of a nested list of arbitrary depth is to [define the type recursively](https://stackoverflow.com/a/53845083/3154101), but [mypy does not currently support recursive type definitions](https://github.com/python/mypy/issues/731), though some other type checkers can, e.g. [Microsoft's Pylance does](https://devblogs.microsoft.com/python/pylance-introduces-five-new-features-that-enable-type-magic-for-python-developers/). We're going to have the same problem when specifying types for `open_mfdataset`. For now this problem is just ignored by the type checker, meaning that we don't actually check the type of the nested-list-of-lists.

- [x] Passes `pre-commit run --all-files`
- [ ] ~~User visible changes (including notable bug fixes) are documented in `whats-new.rst`~~

### CODE REVIEW

File: xarray/core/combine.py [hunk 1]
```diff
@@ -353,16 +356,23 @@ def _nested_combine(
     return combined
 
 
+# Define type for arbitrarily-nested list of lists recursively
+# Currently mypy cannot handle this but other linters can (https://stackoverflow.com/a/53845083/3154101)
+DATASET_HYPERCUBE = Union[Dataset, Iterable["DATASET_HYPERCUBE"]]  # type: ignore
```
Comment (TomNicholas) [inline xarray/core/combine.py:361]: The weird recursive definition is here

**Comment by TomNicholas** [comment]
(It does pass `pre-commit` locally, the linting failures in the CI are due to #5517)

**Comment by max-sixty** [comment]
Thanks @TomNicholas !

**Comment by max-sixty** [comment]
@TomNicholas given the upstream changes, should we merge main in first and then merge?

**Pull request merged**